staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -1322,9 +1357,18 @@ func (v *unstructuredSchemaCoercer) apply(u *unstructured.Unstructured) error {
 		return err
 	}
 	if gv.Group == v.structuralSchemaGK.Group && kind == v.structuralSchemaGK.Kind {
-		if !v.preserveUnknownFields {
+		if v.unknownFieldsDirective != preserve {
 			// TODO: switch over pruning and coercing at the root to  schemaobjectmeta.Coerce too
+			// TODO: this seems like a very error prone way of detecting unknown fields
+			// copy the original object to detect whether any fields are pruned
+			objCopy := (&unstructured.Unstructured{
+				Object: u.Object,
+			}).DeepCopy()
 			structuralpruning.Prune(u.Object, v.structuralSchemas[gv.Version], false)
```

Comment:
can't we add some bool return value to the pruning algorithm?

I don't like the deepcopy plus deepequal here.

Reply:
at the very least it would have to skip the deepcopy when it is not needed